### ISSUE TITLE
Improving the flow of iterative updates - Experiment with Post Merge Reviews (?)

### ISSUE THREAD

**Issue by DanVanAtta** [triplea-game/triplea#5644]
I've been thinking about this for some time, I think it's maybe time we try post-merge reviews. We should formalize it first before we give it a try.

### Givens:
- We know we want updates to be small in size and incremental
- CR is a bottleneck, it is blocking and depends on round-trip communication cycles. Round-trip communication cycles can be a day or more, eg: 3 iterations, can be 3 or more days of being blocked.
- Some updates require many iterations, 3, 5, sometimes more
- There are some very legacy parts of TripleA  we want to clean up, these areas need *many* iterations. The round-trip delay on each iteration makes it infeasible to accomplish these clean-ups. 

### Problems:
- Round trip delays encourage us to pack more in each iteration, increasing the size of PR.
- Splitting up PRs after the fact is time intensive
- We have very few reviewers, we are not full time on this project
- We do not give expectations of how long a review will take, an iterative PR might get lucky and be merged in a few hours, or a few days. This makes it hard to plan iterations and how large each one should be to best use those times of delay. IE: if I can add more to a PR, increasing review time by 15 minutes, and still get the same turn-around time, then I'll do that. Some times though it's better to not add the extra scope if the review turn-around time will be hours.
- We can't cleanup the worst code because it requires too many small iterations that incur too much time, we still have to make forward progress, so we instead we add to the mess. Let's say a method rename with an update as an example, instead of renaming a method and then making changes, we either create a large PR with the method name change and the updates, or we just do not update the method name at all.

### Assumptions:
- we are attempting to do peer review and not gatekeeping
  - This means we do not need *the* 'expert' nor *the* 'owner' to sign off on a change
  - If two people agree, it's probably okay, if not, dissent can be taken up in an issue and explored
If we are doing gatekeeping reviews, it's okay (maybe), but we should be clear about it. I've opened #5643 to explore this topic in further depth to clarify why we are reviewing.

### Proposal:

Post-merge PRs. 

This only applies to those of us with write access. For changes that are pretty clear, minimal, well scoped, we merge them and get reviews post-merge. I think this could allow us to move much faster. For example, method renames could be queued up first, merged, and then the changes that build on it can come later. If we are still reviewing merged code, we are still reviewing all code. We will need to ensure that we can trust that even post-merge comments are followed up on. A post-merge review probably should bias more heavily to following up on all commments than we would for a pre-merge review, or we would lose that trust.

#### What would the mechanics be?

The changes that would be eligible should be single-focus, relatively small. After submission, we let the build pass, merge, and add a "post-merge review" label. As reviewers, we filter by closed and 'post-merge review' to see which items should be reviewed. If we start a conversation, we add 'discussion requested' label to the PR and assign to the author and any participants in the conversation. If we want something to change, we add a 'follow-up requested' label and assign to the author.

Example A) No feedback
- a PR renaming a method is opened
- build passes, a 'post-merge review' label is added
- the PR is merged
- a reviewer opens the merged PR, reviews it
- reviewer sees no issues, they remove teh 'post-merge review' label (or we could add a 'reviewed' label if we want to keep the tracking of which PRs are pre and which are post-merged)

Example B) Conversation feedback
- a PR is opened to fix a bug with a one-line change
- build passes, a 'post-merge review' label is added
- the PR is merged
- a reviewer opens the merged PR, reviews it
- the reviewer has a question about the fix, perhaps if all cases have been considered
- reviewer assigns the PR to author and themself and adds "conversation needed" label
- author replies:
   - follow-up happens immediately with additional pre-merge reviewed PRs
   - no follow-up is needed, questions are answered
   - follow-up is not immediate, we assign to the author, and add a 'follow-up' needed label.
     - if the follow-up is not the cause/fault of the author and just needs to be done, we create an issue with 'volunteers needed' label

Example C) Follow-up requested

- a PR is opened to fix a bug with a one-line change
- build passes, a 'post-merge review' label is added
- the PR is merged
- a reviewer opens the merged PR, reviews it
- the reviewer thinks something should change
- reviewer adds 'follow-up' label, assigns to the author only
 

 ### Concrete Example:

The below diff shows an update that is in-flight. The top of the diff is a critical bug-fix, the bottom is a necessary update for the in-flight change. There is potential for merge conflict in this file and the test-cases if this is done as two steps. The bug-fix iterations add a delay of a day or two or more. The in-flight work meanwhile needs to wait, creating a long-lived branch problem and overhead to resolve merge conflicts if we merge the bug fix change back into the larger branch.

The incentive here is wrong, as the incentive is to include this bug-fix with other changes. If we could instead do a post-merge review of the bug-fix, then these problems go away:
 - the bug-fix is deployed and merged within hours\
 - the larger branch can be rebased onto the bug fix for no conflicts post-merge
 - the larger branch does not have to wait for days before it can be submitted
 - the long-lived branch problem is solved
 - the larger branch does not have the same level of conflict potential, there can easily be 'other' changes that come through and cause conflicts while the brnach is waiting
 - context is kept about the large in-flight branch and it does not have to wait on the side-lines for a few days
 - using time 'available' now is critical for us. We can often have time today, tomorrow, but not the day after. Burning available time waiting for round-trip reviews IMO is one of the ways we lose and this project falters.

Below is an example of the diff, this is just one file of 70 that are changed in the larger branch, this gives an idea of how many different 'carve-ups' are needed and iterations needed, and this is just one of a dozen features that are needed for a larger overall feature (Project#12, http-server):

![Screenshot from 2019-12-02 13-59-38](https://user-images.githubusercontent.com/12397753/70002216-31f6ab80-1514-11ea-95d5-87a4c92018da.png)

In the above screenshot, note the parens, those are critical, we have a major bug without them.

**Comment by DanVanAtta** [comment]
A quick follow-up thought, post-review PRs we should probably try to keep open as long as we can and still do a best effort to get things reviewed pre-merge. So for example if we have a building block PR, and the 'main' branch is not yet ready, then we'd leave the building block PR open. When the 'main' branch is ready to be submitted, then we'd opt for the post-merge review.

**Comment by RoiEXLab** [comment]
I could live with your second suggestion.

**Comment by DanVanAtta** [comment]
To clarify, I'm thinking of it as one suggestions, just saying we won't be eager to go for the post-merge review unless it helps clear the way for another PR to be opened.

**Comment by ron-murhammer** [comment]
I'm open to the idea though it needs to really only be used when folks are looking to follow up on that same code area and that the change is either new feature that has community consensus or just refactoring (and preferably simple). Any user facing changes that don't have strong agreement, things that impact rules/AI, or bug fixes should be reviewed before they are merged to avoid needing reverts.

**Comment by DanVanAtta** [comment]
I suspect we should discuss and formalize how changes can be made in ares of broad interest. Getting feedback on forums is useful, getting concensus though between 5-7 opinionated people at differing priorities and timeframes is ineffective. That is a different topic though and we perhaps should explore it in another thread to help eliminate that friction.

To clarify for this topic, the intent is so that preliminary changes can be submitted with more confidence. It was a bit of a long proposal, some things are detailed.. Please let me know if the guidelines of when such a process should be used are actually not clear, happy to clarify further/answer questions.

**Comment by prastle** [comment]
Off topic but yes the players = forums input is usefull including the mods etc that do not use Git!  Sorry small giggle poke dan  :)

**Comment by DanVanAtta** [comment]
No, I agree @prastle on feedback, but making timely decisions via that process has not been effective. Let's not open that discussion here on this topic.

**Comment by DanVanAtta** [comment]
Okay, I think we're on the same page for this policy. Notably, post-merge review is more or less a tool of least preference so long as it enables us to iterate and not lob giant PRs up for review.

**Comment by DanVanAtta** [comment]
Re-opening, I would like to rescind this policy. I don't think it has worked out. We need to IMO instead better balance when one of us is able to spend a couple days on TripleA and output a burst of work. In the meantime commit-stacking in PRs seems to be a better approach, and always we probably need to just prioritize review over just about any other type of activity. The big problem we see is really the blocker of waiting for review, if we fix that, then commit-stacking and slow PR iteration problems go away.

**Comment by DanVanAtta** [comment]
cc: @ron-murhammer @RoiEXLab

**Comment by ron-murhammer** [comment]
@DanVanAtta Not following exactly what you'd like to rescind. I haven't really seen us have a big backlog of PRs. The bigger issue seems to be getting to a stable point so we can try to push a release as its been over a year.

**Comment by DanVanAtta** [comment]
When I was able to work on TripleA full-time for some time, this was a killer problem. The backlog was not put up as many updates built on each other and I was waiting for reviews and PRs before the next PR to land was ready. In one case a series of 3 PRs took about 3 weeks before the last one could be opened. One days worth of work, took a month to get merged due to turn-around times.

Recently I've been stacking commits and have less bandwidth, so the problem has more or less gone away, and hence there is not much need to merge items for post-merge review. On a similar token, a premise of the post-merge review is that we do not review commit-by-commit well, which is something I think we can simply do instead of post-merge review (which I don't think we've done either).

**Comment by RoiEXLab** [comment]
Good Point @DanVanAtta 
But as already mentioned I haven't seen us relying on post-merge reviews either recently, so all good.

**Comment by DanVanAtta** [comment]
I did observe we did not remove or even come back to PRs with post-merge label. It did not seem to be working. In part, @RoiEXLab you went above and beyond reviewing more than any one person should have. It was great, but it's a trap too as it's not sustainable and not necessarily a global benefit to everyone. It's an ironic situation to be in, I put myself once there, and though it was very ironic that the response to "you're all welcome, see how nobody was blocked with reviews!" was a no-thank you as the reviews were not always careful enough and did not always reflect team consensus or help the full time get knowledge of what was changing.

With that aside, commit-stacking seems to be working okay, it can run into trouble when changes start to sit in-flight and then accumulate conflicts between each other. With a lower cadence of change incoming, it's less of a problem. I suspect it comes down to, if someone ever gets to spend 50 hours in a week on TripleA, perhaps post-merge reviews is the answer and have them self-comment and put up very fine-grained PRs. The situation has changed since then though. The only problems I'm seeing is that commit stacking might not work if we don't click through commits, answering questions where we took time to clearly spell out the reasons in a commit message can be a bit frustrating.
